src/notify.rs

```diff
@@ -0,0 +1,146 @@
+use std::cell::Cell;
+use std::cell::RefCell;
+use std::collections::VecDeque;
+use std::future::Future;
+use std::pin::Pin;
+use std::rc::Rc;
+use std::task::Context;
+use std::task::Poll;
+use std::task::Waker;
+
+#[derive(Default, Copy, Clone, PartialEq, Eq)]
+enum NotifyState {
+  #[default]
+  Pending,
+  Notified,
+  Dropped,
+}
+
+struct WakerWithState {
+  inner: Waker,
+  notify_state: Rc<Cell<NotifyState>>,
+}
+
+impl WakerWithState {
+  pub fn notify(self) {
+    debug_assert!(self.notify_state.get() == NotifyState::Pending);
+    self.notify_state.set(NotifyState::Notified);
+    self.inner.wake();
+  }
+
+  pub fn is_future_dropped(&self) -> bool {
+    self.notify_state.get() == NotifyState::Dropped
+  }
+}
+
+#[derive(Default)]
+pub struct NotifyInner {
+  pending_notify: usize,
+  wakers: VecDeque<WakerWithState>,
+}
+
+#[derive(Default)]
+pub struct Notify(Rc<RefCell<NotifyInner>>);
+
+impl Notify {
+  pub fn new() -> Self {
+    Self::default()
+  }
+
+  pub fn notify_one(&self) {
+    let mut maybe_waker = None;
+    {
+      let mut inner = self.0.borrow_mut();
+      while let Some(waker) = inner.wakers.pop_front() {
+        if !waker.is_future_dropped() {
+          maybe_waker = Some(waker);
+          break;
+        }
+      }
+      // no wakers, so make the next one be notified immediately
+      if maybe_waker.is_none() {
+        inner.pending_notify += 1;
+      }
+    }
+    if let Some(waker) = maybe_waker {
+      waker.notify();
+    }
+  }
+
+  pub fn notified(&self) -> impl Future<Output = ()> {
+    NotifiedFuture {
+      inner: self.0.clone(),
+      notify_state: Default::default(),
+    }
+  }
```

Comment:
Nothing really happens with the internal state of `Notify` until the future is polled. Is that ok?